Fix the recursively included issue for srs_core_time.hpp

mingo-wu1 opened the pull request:
Please describe the summary for this PR.

1. In included file: main file cannot be included recursively when building a preamble
    clang(pp_including_mainfile_in_preamble)
    srs_core.hpp(43, 10): Error occurred here
    Circular reference to header files

---------

`TRANS_BY_GPT3`

mingo-wu1 committed:
Update srs_core_time.hpp

In included file: main file cannot be included recursively when building a preambleclang(pp_including_mainfile_in_preamble)
srs_core.hpp(43, 10): Error occurred here
循环引用头文件

winlinvip commented on trunk/src/core/srs_core_time.hpp:
```diff
@@ -7,10 +7,8 @@
-
 // Time and duration unit, in us.
-typedef int64_t srs_utime_t;
+typedef signed long int srs_utime_t;
```
Why change this? Not causes the recursively including.

mingo-wu1 replied:
Here is a detailed analysis to indicate that I have a lack of thinking (please see in detail, I will make more rigorous changes):

1. int64_t is defined on my machine Ubuntu 20.04 Focal at /usr/include/x86_64-linux-gnu/bits/stdint-intn.h, line 27, with the code typedef __int64_t int64_t; If int64_t is needed, the header file <bits/stdint-intn.h> or a higher-level header file needs to be included.

2. When removing the inclusion of the header file <srs_core.hpp> due to header file circular referencing, the trunk/src/core/srs_core_time.hpp does not include any header files. At this point, int64_t will throw an error indicating that the definition cannot be found (this error will occur with both gcc and g++).

3. At this point, there are two solutions: <1> include the header files included in the header file or <2> modify int64_t to signed long int (on my machine, int64_t is defined by signed long int, but there is a lack of consideration here).

4. After my more rigorous analysis, I believe that replacing typedef int64_t srs_utime_t with typedef signed long int srs_utime_t is inadequate because I only considered my own machine.

5. Because the rule exists that short <= int <= long <= long long, and in some cases, the size of long may not be 64 bits, such as on a 32-bit machine where long is 32 bits.

6. I checked the IDEs of Visual Studio on Windows and VSCode on Ubuntu (the versions are not relevant here), and int64_t is included in <stdint.h>.

7. Therefore, I referred to the usage of the header file <stdint.h> in the srs project in trunk/src/app/srs_app_rtc_sdp.hpp, trunk/src/app/srs_app_uuid_cpp, and trunk/src/protocol/srs_protocol_http_stack.hpp.
<1>  trunk/src/app/srs_app_rtc_sdp.hpp
#include <stdint.h>
<2>   trunk/src/app/srs_app_uuid_cpp
#ifdef HAVE_STDINT_H
#include <stdint.h>
#else
#ifdef HAVE_INTTYPES_H
#include <inttypes.h>
#endif
#endif
<3>  trunk/src/protocol/srs_protocol_http_stack.hpp
#if defined(_WIN32) && !defined(__MINGW32__) && \
  (!defined(_MSC_VER) || _MSC_VER<1600) && !defined(__WINE__)
#include <BaseTsd.h>
typedef __int8 int8_t;
typedef unsigned __int8 uint8_t;
typedef __int16 int16_t;
typedef unsigned __int16 uint16_t;
typedef __int32 int32_t;
typedef unsigned __int32 uint32_t;
typedef __int64 int64_t;
typedef unsigned __int64 uint64_t;
#else
#include <stdint.h>
#endif

It can be inferred that some environments do not include <stdint.h>.

8. So I used gcc and g++ to create simple test files to test the above three referencing strategies: gcc -o test test.cpp and g++ -o test test.cpp (the second one will throw an error: unterminated). Therefore, I believe a more rigorous approach is to consider the system environment more comprehensively.
#if defined(_WIN32) && !defined(__MINGW32__) && \
  (!defined(_MSC_VER) || _MSC_VER<1600) && !defined(__WINE__)
#include <BaseTsd.h>
typedef __int64 int64_t;
#else
#include <stdint.h>
#endif

9. Summary:
So far, we have considered the operating system's bitness, different header file dependencies of the operating system, and the IDE. After conducting compilation tests, the inclusion of int64_t should be done using the following approach: adopting the strategy of trunk/src/protocol/srs_protocol_http_stack.hpp and making modifications to it.
#if defined(_WIN32) && !defined(__MINGW32__) && (!defined(_MSC_VER) || _MSC_VER<1600) && !defined(__WINE__)
#include <BaseTsd.h>
typedef __int64 int64_t;
#else
#include <stdint.h>
#endif

If you have any doubts, please raise them. It is possible that any oversight on my part may have caused them.

`TRANS_BY_GPT3`

winlinvip replied:
Good job! Thanks.

mingo-wu1 committed:
Update srs_core_time.hpp

这里做个详细分析表示本人有思考欠缺(请详看,我会做更严谨的更改)
1. int64_t 在本人机器 ubuntu20.04 focal 定义在 /usr/include/x86_64-linux-gnu/bits/stdint-intn.h , 27行代码 typedef __int64_t int64_t ; 如果需要使用 int64_t 那么需要引入头文件 <bits/stdint-intn.h> 或者更上层头文件
2. 当因为头文件循环引用问题去掉 include <srs_core.hpp> 这个头文件的时候
    trunk/src/core/srs_core_time.hpp 是没有引入任何的头文件的, 此时 int64_t 就会报错找不到定义的地方(使用gcc或者g++都会报错)
3. 此时这里有两个解决办法就是 <1> 引入头文件包含的头文件 或者 <2> 修改 int64_t 为 signed long int (这里本人机器int64_t 是由 signed long int 定义的, 这里有考虑欠缺)
4. 经过我更严谨的分析,我觉得我的这个 typedef int64_t srs_utime_t 替换成 typedef signed long int srs_utime_t 办法有欠缺, 因为我只考虑了本人机器,
5. 因为存在规则 short <= int <= long <= long long , 并且存在一些情况 long 的大小可能不是64位, 比如在32位机器下long是32位
6. 我查看了win下 visual studio 的 IDE 和 ubuntu 下 vscode 的 IDE(这里无需考虑版本) , int64_t 被包含在 <stdint.h>
7. 于是我参考了srs项目中 trunk/src/app/srs_app_rtc_sdp.hpp , trunk/src/app/srs_app_uuid_cpp, trunk/src/protocol/srs_protocol_http_stack.hpp 中使用头文件 <stdint.h> 的情况:
<1>  trunk/src/app/srs_app_rtc_sdp.hpp
#include <stdint.h>
<2>   trunk/src/app/srs_app_uuid_cpp
#ifdef HAVE_STDINT_H
#include <stdint.h>
#else
#ifdef HAVE_INTTYPES_H
#include <inttypes.h>
#endif
#endif
<3>  trunk/src/protocol/srs_protocol_http_stack.hpp
#if defined(_WIN32) && !defined(__MINGW32__) && \
  (!defined(_MSC_VER) || _MSC_VER<1600) && !defined(__WINE__)
#include <BaseTsd.h>
typedef __int8 int8_t;
typedef unsigned __int8 uint8_t;
typedef __int16 int16_t;
typedef unsigned __int16 uint16_t;
typedef __int32 int32_t;
typedef unsigned __int32 uint32_t;
typedef __int64 int64_t;
typedef unsigned __int64 uint64_t;
#else
#include <stdint.h>
#endif

推理得出有的环境没有包含<stdint.h>

8. 于是我用gcc和g++建立简单的测试文件进行上面三种引用策略的测试 gcc -o test test.cpp 和 g++ -o test test.cpp (其中第二种会报错因为error: unterminated), 所以认为更严谨的处理办法是(更周全的考虑的系统环境):
#if defined(_WIN32) && !defined(__MINGW32__) && \
  (!defined(_MSC_VER) || _MSC_VER<1600) && !defined(__WINE__)
#include <BaseTsd.h>
typedef __int64 int64_t;
#else
#include <stdint.h>
#endif

9. 总结:
至此, 考虑了操作系统的位数, 考虑了操作系统的不同头文件依赖, 也考虑了IDE,  同时查看过 BaseTsd.h
int64_t 的包含采用下面方式,
采用 trunk/src/protocol/srs_protocol_http_stack.hpp 的策略并进行修改为:
#if defined(_WIN32) && !defined(__MINGW32__) && (!defined(_MSC_VER) || _MSC_VER<1600) && !defined(__WINE__)
#include <BaseTsd.h>
typedef __int64 int64_t;
#else
#include <stdint.h>
#endif

如果有疑惑请提出, 可能会因为本人考虑不周引起的.

winlinvip commented:
I make a small change, directly define the srs_utine_t, do not define int64_t, because it might cause redefine issue.

```
// Time and duration unit, in us.
#if defined(_WIN32) && !defined(__MINGW32__) && (!defined(_MSC_VER) || _MSC_VER<1600) && !defined(__WINE__)
#include <BaseTsd.h>
typedef __int64 srs_utime_t;
#else
#include <stdint.h>
typedef int64_t srs_utime_t;
#endif
```
